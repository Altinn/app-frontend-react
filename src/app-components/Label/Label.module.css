.label {
  padding-bottom: var(--fds-spacing-2);
  display: flex;
  gap: var(--fds-spacing-1);
}
