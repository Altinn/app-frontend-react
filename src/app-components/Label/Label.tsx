import React from 'react';
import type { JSX, PropsWithChildren, ReactElement } from 'react';

import { Label as DesignsystemetLabel } from '@digdir/designsystemet-react';
import { Grid } from '@material-ui/core';
import cn from 'classnames';
import type { LabelProps as DesignsystemetLabelProps } from '@digdir/designsystemet-react';

import classes from 'src/app-components/Label/Label.module.css';
import type { GridSize } from 'src/app-components/Label/types';

type LabelProps = {
  label: string | undefined;
  optionalIndicator?: ReactElement;
  help?: ReactElement;
  description?: ReactElement;
  className?: string;
  grid?: GridSize;
  required?: boolean;
  requiredIndicator?: JSX.Element;
  htmlFor?: DesignsystemetLabelProps['htmlFor'];
  style?: DesignsystemetLabelProps['style'];
};

export function Label({
  label,
  required,
  requiredIndicator,
  optionalIndicator,
  help,
  description,
  htmlFor,
  style,
  className,
  grid,
  children,
}: PropsWithChildren<LabelProps>) {
  if (!label) {
    return children;
  }

  return (
    <Grid
      container
      spacing={2}
    >
      <Grid
        item
        {...(grid ?? { xs: 12 })}
      >
        <span className={classes.labelAndDescWrapper}>
          <span className={classes.labelAndHelpWrapper}>
            <DesignsystemetLabel
              weight='medium'
              size='md'
              htmlFor={htmlFor}
              className={cn(className)}
              style={style}
            >
              {label}
              {required && requiredIndicator}
              {!required && optionalIndicator}
            </DesignsystemetLabel>
            {help}
          </span>
          {description}
        </span>
      </Grid>
      {children}
    </Grid>
  );
}
