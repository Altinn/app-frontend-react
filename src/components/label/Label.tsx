import React from 'react';
import type { PropsWithChildren } from 'react';

import { Fieldset, Label as DesignsystemetLabel } from '@digdir/designsystemet-react';
import { Grid } from '@material-ui/core';

import classes from 'src/components/label/Label.module.css';
import { LabelContent } from 'src/components/label/LabelContent';
import { gridBreakpoints } from 'src/utils/formComponentUtils';
import type { LabelContentProps } from 'src/components/label/LabelContent';
import type { IGridStyling, ILabelSettings } from 'src/layout/common.generated';

type LabelType = 'legend' | 'span' | 'label';

export type LabelProps = PropsWithChildren<{
  id: string;
  renderLabelAs: LabelType;
  required?: boolean;
  readOnly?: boolean;
  labelSettings?: ILabelSettings;
  grid?: { labelGrid?: IGridStyling };
  textResourceBindings?: {
    title?: string;
    description?: string;
    helpText?: string;
  };
}>;

export function Label({
  id,
  renderLabelAs,
  required,
  readOnly,
  labelSettings,
  grid,
  textResourceBindings,
  children,
}: LabelProps) {
  if (!textResourceBindings?.title) {
    return <>{children}</>;
  }

  const labelId = `label-${id}`;
  const labelContentProps: Omit<LabelContentProps, 'id'> = {
    label: textResourceBindings.title,
    description: textResourceBindings.description,
    helpText: textResourceBindings.helpText,
    required,
    readOnly,
    labelSettings,
  };

  switch (renderLabelAs) {
    case 'legend': {
      return (
        <Fieldset
          size='small'
          className={classes.fieldWrapper}
          legend={
            <Grid {...gridBreakpoints(grid?.labelGrid)}>
              <LabelContent
                id={labelId}
                {...labelContentProps}
              />
            </Grid>
          }
        >
          {children}
        </Fieldset>
      );
    }
    case 'label':
      return (
        <DesignsystemetLabel
          id={labelId}
          htmlFor={id}
          style={{ width: '100%' }}
        >
          <Grid container>
            <Grid {...gridBreakpoints(grid?.labelGrid)}>
              <LabelContent {...labelContentProps} />
            </Grid>
            {children}
          </Grid>
        </DesignsystemetLabel>
      );

    case 'span':
    default:
      return (
        <Grid container>
          {/* we want this "label" not to be rendered as a <label>,
           because it does not belong to an input element */}
          <Grid {...gridBreakpoints(grid?.labelGrid)}>
            <DesignsystemetLabel asChild>
              <LabelContent
                id={labelId}
                {...labelContentProps}
              />
            </DesignsystemetLabel>
          </Grid>
          {children}
        </Grid>
      );
  }
}
