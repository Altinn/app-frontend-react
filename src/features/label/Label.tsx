import React from 'react';
import type { PropsWithChildren } from 'react';

import { Fieldset, Label as DesignsystemetLabel } from '@digdir/designsystemet-react';

import classes from 'src/features/label/Label.module.css';
import { LabelContent } from 'src/features/label/LabelContent';
import type { LabelContentProps } from 'src/features/label/LabelContent';
import type { ILabelSettings } from 'src/layout/common.generated';

type LabelProps = PropsWithChildren<{
  id: string;
  renderLabelAs: 'legend' | 'span' | 'label';
  required?: boolean;
  readOnly?: boolean;
  labelSettings?: ILabelSettings;
  textResourceBindings?: {
    title?: string;
    description?: string;
    helpText?: string;
  };
}>;

export function Label({
  id,
  renderLabelAs,
  children,
  textResourceBindings,
  required,
  readOnly,
  labelSettings,
}: LabelProps) {
  if (!textResourceBindings?.title) {
    return <>{children}</>;
  }

  const labelId = `label-${id}`;
  const labelContentProps: Omit<LabelContentProps, 'id'> = {
    label: textResourceBindings.title,
    description: textResourceBindings.description,
    helpText: textResourceBindings.helpText,
    required,
    readOnly,
    labelSettings,
  };

  switch (renderLabelAs) {
    case 'legend': {
      return (
        <Fieldset
          size='small'
          className={classes.fieldWrapper}
          legend={
            <LabelContent
              id={labelId}
              {...labelContentProps}
            />
          }
        >
          {children}
        </Fieldset>
      );
    }
    case 'label':
      return (
        <div className={classes.fieldWrapper}>
          <DesignsystemetLabel
            id={labelId}
            htmlFor={id}
          >
            <LabelContent {...labelContentProps} />
          </DesignsystemetLabel>
          {children}
        </div>
      );

    case 'span':
    default:
      return (
        <div className={classes.fieldWrapper}>
          {/* we want this "label" to be rendered as a <span> and not a <label>,
           because it does not belong to an input element */}
          <DesignsystemetLabel asChild>
            <LabelContent
              id={labelId}
              {...labelContentProps}
            />
          </DesignsystemetLabel>
          {children}
        </div>
      );
  }
}
