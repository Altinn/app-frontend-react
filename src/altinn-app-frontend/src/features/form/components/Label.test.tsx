import * as React from 'react';
import { render } from '@testing-library/react';

import Label from './Label';
import type { IFormLabelProps } from './Label';

describe('features > form > components >Label.tsx', () => {
  const requiredMarking = '*';
  const optionalMarking = 'Valgfri';

  function renderLabelComponent(props: Partial<IFormLabelProps> = {}) {
    const defaultProps: IFormLabelProps = {
      id: 'label1',
      labelText: 'label.text',
      helpText: '',
      language: {
        general: {
          optional: optionalMarking,
        },
        'form_filler': {
          'required_label': requiredMarking,
        },
      },
      readOnly: false,
      required: false,
      labelSettings: {
        optionalIndicator: true,
      }
    };

    return render(<Label {...defaultProps} {...props} />);
  }

  it('should render label', () => {
    const { queryByText } = renderLabelComponent();
    expect(queryByText('label.text')).toBeInTheDocument();
  })

  it('should render required marking when field is required', () => {
    const { queryByText } = renderLabelComponent({ required: true });
    expect(queryByText(requiredMarking)).toBeTruthy();
  });

  it('should not render required marking when field is readOnly', () => {
    const { queryByText } = renderLabelComponent({ required: true, readOnly: true });
    expect(queryByText(requiredMarking)).toBeFalsy();
  });

  it('should render optional marking when labelSettings.optionalIndicator is true, and required, readOnly are both false', () => {
    const { queryByText } = renderLabelComponent({ labelSettings: { optionalIndicator: true } });
    expect(queryByText(`(${optionalMarking})`)).toBeTruthy();
  });

  it('should not render optional marking when required, even if labelSettings.optionalIndicator is true', () => {
    const { queryByText } = renderLabelComponent({ labelSettings: { optionalIndicator: true }, required: true });
    expect(queryByText(` (${optionalMarking})`)).toBeFalsy();
  });

  it('should not render optional marking when readOnly, even if labelSettings.optionalIndicator is true', () => {
    const { queryByText } = renderLabelComponent({ labelSettings: { optionalIndicator: true }, readOnly: true });
    expect(queryByText(` (${optionalMarking})`)).toBeFalsy();
  });
});
