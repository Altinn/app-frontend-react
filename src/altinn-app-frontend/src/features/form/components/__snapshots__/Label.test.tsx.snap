// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`features > form > components >Label.tsx should render default 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12"
  >
    <div
      class="MuiGrid-root MuiGrid-item"
    >
      <label
        class="a-form-label title-label"
        data-testid="label-label1"
        for="label1"
      >
        label.text
      </label>
    </div>
  </div>
</DocumentFragment>
`;
